### PR TITLE
ECC: sign and self tests, at crypto_ops level

### PR DESCRIPTION
**Pull request by pascal-brand-st-dev**
Signed-off-by: Pascal Brand <pascal.brand@st.com>

### CODE REVIEW

File: core/lib/libtomcrypt/include/tomcrypt_pk.h [hunk 1]
```diff
@@ -288,10 +288,18 @@ int  ecc_decrypt_key(const unsigned char *in,  unsigned long  inlen,
                            unsigned char *out, unsigned long *outlen, 
                            ecc_key *key);
 
+int ecc_sign_hash_raw(const unsigned char *in,  unsigned long inlen,
+                            void   *r,   void *s,
+                            prng_state *prng, int wprng, ecc_key *key);
+
```
Comment (jforissier) [inline core/lib/libtomcrypt/include/tomcrypt_pk.h]: Weird spacing

File: core/lib/libtomcrypt/include/tomcrypt_pk.h [hunk 2]
```diff
@@ -288,10 +288,18 @@ int  ecc_decrypt_key(const unsigned char *in,  unsigned long  inlen,
...
 int  ecc_sign_hash(const unsigned char *in,  unsigned long inlen, 
                          unsigned char *out, unsigned long *outlen, 
                          prng_state *prng, int wprng, ecc_key *key);
 
+int ecc_verify_hash_raw(      void   *r, void   *s,
+                        const unsigned char *hash, unsigned long hashlen,
+                        int *stat, ecc_key *key);
+
```
Comment (jforissier) [inline core/lib/libtomcrypt/include/tomcrypt_pk.h]: Same as above

File: core/lib/libtomcrypt/src/pk/ecc/ecc_verify_hash.c [hunk 3]
```diff
@@ -178,13 +171,57 @@ int ecc_verify_hash(const unsigned char *sig,  unsigned long siglen,
...
+   @param hash        The hash (message digest) that was signed
+   @param hashlen     The length of the hash (octets)
+   @param stat        Result of signature, 1==valid, 0==invalid
+   @param key         The corresponding public ECC key
+   @return CRYPT_OK if successful (even if the signature is not valid)
+*/
+
+int ecc_verify_hash(const unsigned char *sig,  unsigned long siglen,
+                    const unsigned char *hash, unsigned long hashlen,
+                    int *stat, ecc_key *key)
```
Comment (jforissier) [inline core/lib/libtomcrypt/src/pk/ecc/ecc_verify_hash.c]: I think `stat` is not needed, just return an error status instead. I.e., `CRYPT_OK` means signature could be verified and is OK, any other code means it's not or it could not be verified.
I see other functions are like this one (`dsa_verify_hash_raw()` for instance) so it could be the purpose of another patch if you think this comment is valid.

File: core/lib/libtomcrypt/src/pk/ecc/ltc_ecc_mulmod.c [hunk 4]
```diff
@@ -46,13 +46,13 @@
...
   ECC Crypto, Tom St Denis
-*/  
+*/
 
 #ifdef LTC_MECC
 #ifndef LTC_ECC_TIMING_RESISTANT
 
 /* size of sliding window, don't change this! MUST BE A POWER OF TWO */
-#define WINSIZE (2)
+#define WINSIZE 4
```
Comment (jforissier) [inline core/lib/libtomcrypt/src/pk/ecc/ltc_ecc_mulmod.c]: Hum... why did you change it? Is it performance-related?

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 5]
```diff
@@ -465,7 +465,11 @@ static TEE_Result hash_final(void *ctx, uint32_t algo, uint8_t *digest,
 #if defined(_CFG_CRYPTO_WITH_ACIPHER)
 
 #define LTC_MAX_BITS_PER_VARIABLE   (4096)
+#ifdef CFG_CRYPTO_ECC
+#define LTC_VARIABLE_NUMBER         (80)
+#else
 #define LTC_VARIABLE_NUMBER         (50)
+#endif
```
Comment (jforissier) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: We need more than the default?

File: core/sta/ecc_self_tests.c [hunk 6]
```diff
@@ -208,6 +373,8 @@ static TEE_Result invoke_command(void *pSessionContext __unused,
 	switch (nCommandID) {
 	case CMD_ECC_GEN_KEY_SELF_TESTS:
 		return ecc_generate_key_tests(pParams[0].value.a);
+	case CMD_ECC_DSA_TESTS:
+		return ecc_dsa_tests(pParams[0].value.a);
```
Comment (jforissier) [inline core/sta/ecc_self_tests.c]: I'm not a big fan of having tests in optee_os, they belong to optee_test IMO. The problem here is that you don't provide the client application that exercises the self-tests (or did I miss something?) so they are of no use to most people.

File: core/lib/libtomcrypt/src/tee_ltc_provider.c (hunk 5 above)
Comment (pascal-brand-st-dev) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: Yes, ECC requires lots of variables. 50 is not enough.

I think we should think of a way to optimize it because we request, for ECC, 80 temporary variables (which is ok) of size 4096 bits (which is way too more). This could be part of another patch.

File: core/sta/ecc_self_tests.c (hunk 6 above)
Comment (pascal-brand-st-dev) [inline core/sta/ecc_self_tests.c]: This was one comment from Jens on the ECC Gen key. As the patch is at crypto_ops level, we cannot add test the feature at GP level yet (as in xtest 400x). This will be done in further patches. We may then remove this self static TA used for testing crypto_ops.

The client part of the test exist on the linaro-private git, branch ecc_next. This private git will become obsolete soon. So I'll create a pull-request on https://github.com/OP-TEE/optee_test soon.

File: core/lib/libtomcrypt/src/pk/ecc/ltc_ecc_mulmod.c (hunk 4 above)
Comment (pascal-brand-st-dev) [inline core/lib/libtomcrypt/src/pk/ecc/ltc_ecc_mulmod.c]: Libtomcrypt GitHub synchro.

File: core/lib/libtomcrypt/src/pk/ecc/ecc_verify_hash.c (hunk 3 above)
Comment (pascal-brand-st-dev) [inline core/lib/libtomcrypt/src/pk/ecc/ecc_verify_hash.c]: I do not want to change the LibTomCrypt API.

File: core/lib/libtomcrypt/include/tomcrypt_pk.h (hunk 1 above)
Comment (pascal-brand-st-dev) [inline core/lib/libtomcrypt/include/tomcrypt_pk.h]: Yes, I know. I tried to keep the spacing rules (if any) there is on Libtomcrypt, such as dsa_sign_hash_raw(). Let me know if I should use the OPTEE coding rules for this and below remark.

Comment (jforissier) [inline core/lib/libtomcrypt/include/tomcrypt_pk.h]: OK let's keep it as it is for a bit of consistency (and same below of course).

File: core/sta/ecc_self_tests.c (hunk 6 above)
Comment (pascal-brand-st-dev) [inline core/sta/ecc_self_tests.c]: https://github.com/OP-TEE/optee_test/pull/1 created

File: core/lib/libtomcrypt/src/pk/ecc/ecc_verify_hash.c (hunk 3 above)
Comment (jforissier) [inline core/lib/libtomcrypt/src/pk/ecc/ecc_verify_hash.c]: OK, my bad, this is not an OP-TEE private file.

File: core/lib/libtomcrypt/src/pk/ecc/ltc_ecc_mulmod.c (hunk 4 above)
Comment (jforissier) [inline core/lib/libtomcrypt/src/pk/ecc/ltc_ecc_mulmod.c]: Fine.

File: core/sta/ecc_self_tests.c (hunk 6 above)
Comment (jforissier) [inline core/sta/ecc_self_tests.c]: Sounds good.

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey 32 & 64-bit)`

**Comment by jenswi-linaro** [comment]
```
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>
```

**Comment by pascal-brand-st-dev** [comment]
I increase the size of the emulated esram. @jenswi-linaro @jforissier Let me know if it's still ok for you.

**Comment by jforissier** [comment]
@pascal-brand-st-dev yes, you can add Hikey with pager to my Tested-by.

`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey with pager)`

**Comment by jenswi-linaro** [comment]
@pascal-brand-st-dev yes, it's OK.
